espnet/nets/batch_beam_search.py

```diff
@@ -0,0 +1,202 @@
+            scores={k: v[ids] for k, v in hyps.scores.items()},
+            states={k: [self.scorers[k].select_state(v, i) for i in ids]
+                    for k, v in hyps.states.items()},
+        )
+
+    def _select(self, hyps: BatchHypothesis, i: int) -> Hypothesis:
+        return Hypothesis(
+            yseq=hyps.yseq[i, :hyps.length[i]],
+            score=hyps.score[i],
+            scores={k: v[i] for k, v in hyps.scores.items()},
+            states={k: self.scorers[k].select_state(v, i) for k, v in hyps.states.items()},
+        )
+
+    def unbatchfy(self, batch_hyps: BatchHypothesis) -> List[Hypothesis]:
+        """Revert batch to list."""
+        return [
+            Hypothesis(
+                yseq=batch_hyps.yseq[i][:batch_hyps.length[i]],
+                score=batch_hyps.score[i],
+                scores={k: batch_hyps.scores[k][i] for k in self.scorers},
+                states={k: v.select_state(
+                    batch_hyps.states[k], i) for k, v in self.scorers.items()}
+            ) for i in range(len(batch_hyps.length))]
+
+    def batch_beam(self, weighted_scores: torch.Tensor, ids: torch.Tensor) \
+            -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Batch-compute topk full token ids and partial token ids.
+
+        Args:
+            weighted_scores (torch.Tensor): The weighted sum scores for each tokens.
+                Its shape is `(n_beam, self.vocab_size)`.
+            ids (torch.Tensor): The partial token ids to compute topk.
+                Its shape is `(n_beam, self.pre_beam_size)`.
+
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+                The topk full (prev_hyp, new_token) ids and partial (prev_hyp, new_token) ids.
+                Their shapes are all `(self.beam_size,)`
+
+        """
+        if not self.do_pre_beam:
+            top_ids = weighted_scores.view(-1).topk(self.beam_size)[1]
+            # Because of the flatten above, `top_ids` is organized as:
+            # [hyp1 * V + token1, hyp2 * V + token2, ..., hypK * V + tokenK],
+            # where V is `self.n_vocab` and K is `self.beam_size`
+            prev_hyp_ids = top_ids // self.n_vocab
+            new_token_ids = top_ids % self.n_vocab
+            return prev_hyp_ids, new_token_ids, prev_hyp_ids, new_token_ids
+
+        raise NotImplementedError("batch decoding with PartialScorer is not supported yet.")
+
+    def init_hyp(self, x: torch.Tensor) -> BatchHypothesis:
+        """Get an initial hypothesis data.
+
+        Args:
+            x (torch.Tensor): The encoder output feature
+
+        Returns:
+            Hypothesis: The initial hypothesis.
+
+        """
+        init_states = dict()
+        init_scores = dict()
+        for k, d in self.scorers.items():
+            init_states[k] = d.init_state(x)
+            init_scores[k] = 0.0
+        return self.batchfy(super().init_hyp(x))
```

Comment:
Expected type 'List[Hypothesis]', got 'Hypothesis'

Reply:
Thanks. I found that the return type annotation of `BeamSearch.init_hyp` was wrong.